src/polyswarm_api/api.py

```diff
@@ -6,433 +7,382 @@
+        """
+        return self.generator.update_rule_set(rule_set_id, name=name, rules=rules, description=description).execute().result
 
-        return result.HuntResult(hunt, reqs, self)
+    def rule_set_delete(self, rule_set_id):
+        """
+        Delete a YaraRuleset from the polyswarm platform by its Id.
+        :param rule_set_id: Id of the ruleset
+        :return: A YaraRuleset resource
+        """
+        return self.generator.delete_rule_set(rule_set_id).execute().result
 
-    def live_results(self, hunt=None, **kwargs):
+    def rule_set_list(self):
         """
-        Get results from a live hunt
+        List all YaraRulesets for the current account.
+        :return: A generator of YaraRuleset resources
+        """
+        return self.generator.list_rule_set().execute().consume_results()
 
-        :param hunt_id: ID of the hunt (None if latest rule results are desired)
-        :return: HuntResult object
+    def download(self, out_dir, hash_, hash_type=None):
+        """
+        Grab the data of artifact identified by hash, and write the data to a file in the provided directory
+        under a file named after the hash_.
+        :param out_dir: Destination directory to download the file.
+        :param hash_: hash
+        :param hash_type: Hash type of the provided hash_. Will attempt to auto-detect if not explicitly provided.
+        :return: A LocalArtifact resource
         """
-        return self._get_hunt_results(hunt, self.endpoint.live_lookup, **kwargs)
+        hash_ = resources.Hash.from_hashable(hash_, hash_type=hash_type)
+        path = os.path.join(out_dir, hash_.hash)
+        return self.generator.download(hash_.hash, hash_.hash_type, path, create=True).execute().result
 
-    def historical_results(self, hunt=None, **kwargs):
+    def download_archive(self, out_dir, s3_path):
         """
-        Get results from a historical hunt
+        Grab the data in the s3 path provided in the stream() method, and write the contents
+        in the provided directory.
+        :param out_dir: Destination directory to download the file.
+        :param s3_path: Target S3 object to download.
+        :return: A LocalArtifact resource
+        """
+        path = os.path.join(out_dir, os.path.basename(urlparse(s3_path).path))
+        return self.generator.download_archive(s3_path, path, create=True).execute().result
 
-        :param hunt_id: ID of the hunt (None if latest hunt results are desired)
-        :return: HuntResult object
+    def download_to_filehandle(self, hash_, fh, hash_type=None):
         """
-        return self._get_hunt_results(hunt, self.endpoint.historical_lookup, **kwargs)
+        Grab the data of artifact identified by hash, and write the data to a file handle
+        :param hash_: hash
+        :param fh: file handle
+        :param hash_type: Hash type of the provided hash_. Will attempt to auto-detect if not explicitly provided.
+        :return: A LocalArtifact resource
```

Comment:
The `LocalArtifact` is a special resource in the sense that the code there is completely different from the other resources. It was made specifically for downloading files and not jsons. It was meant to represent a binary object in some way and add functionality to this kind of resource, such as running analyzers on the binary, calculating hashes, etc. That said, this particular method `download_to_filehandle` is even more specific, because the user will be providing the handle where it wants the content written, meaning it will likely not be interested in the return value of this method. I am ok with not returning anything in this method (meaning it will return `None`). We could keep a pointer to the handle that `LocalArtifact` used, making it a bit more meaningful, but I agree that does not improve things a lot (actually, if we want to provide functionality such as analyzers for the instance created by `download_to_filehandle`, we would have to keep a pointer to the handle).

I think, however, that the operations we would like to offer on what is returned by `download_to_filehandle` is quite similar to the operations we would like to offer for `download`, `download_archive` and `stream`, that is why I believe they should be the same resource. Downloading files is a special case from the api and I would like to keep the relevant code to deal with this case and the associated relevant methods for the binary in a single resource, even if it is used behind the scenes and not returned. The name "local archive" however is unfortunate, maybe it should just be something like "archive" or "binary", idk. I would like to refrain from creating other resources, specially another special resource that is already non-standard. 

As for contents being available, once the parsing is finished and the object is returned, the contents should have already been written to the handler, so they should be available. I will double check just in case.

Reply:
A file handle does not necessarily retain content, as it could be writable but not readable. For instance, you can use `sys.stdout` here (by design), but certainly can't read that content after it has been written.

I understand the desire to reduce code duplication, but we certainly should not be overloading it to deal with JSONs, binary artifacts, URLs, etc. These all have different operations that would be performed on them, different ways of interacting with them, and only share the common thread that at some point we have the actual content locally. These should be different resources that inherit from a base class imo.

Reply:
Agreed, if the handle is not readable it will not work. I will create a new resource for this, I don't think it is avoiding a lot of code duplication anyways.